hnn_core/tests/test_parallel_backends.py

```diff
@@ -79,65 +93,94 @@ def run_hnn_core(backend=None, n_jobs=1):
-        # account for rounding error incured during MPI parallelization
-        if dpls_reduced_mpi:
+# The purpose of this incremental mark is to avoid running the full length
+# simulation when there are failures in previous (faster) tests. When a test
+# in the sequence fails, all subsequent tests will be marked "xfailed" rather
+# than skipped.
+
+
+@pytest.mark.incremental
+class TestParallelBackends():
+    dpls_reduced_mpi = None
+    dpls_reduced_default = None
+    dpls_reduced_joblib = None
+
+    def test_run_default(self):
+        """Test consistency between default backend simulation and master"""
+        global dpls_reduced_default
+        dpls_reduced_default, _ = run_hnn_core_reduced(None)
+        # test consistency across all parallel backends for multiple trials
+        assert_raises(AssertionError, assert_array_equal,
+                      dpls_reduced_default[0].data['agg'],
+                      dpls_reduced_default[1].data['agg'])
+
+    def test_run_joblibbackend(self):
+        """Test consistency between joblib backend simulation with master"""
+        global dpls_reduced_default, dpls_reduced_joblib
+
+        dpls_reduced_joblib, _ = run_hnn_core_reduced(backend='joblib',
+                                                      n_jobs=2)
+
+        for trial_idx in range(len(dpls_reduced_default)):
+            assert_array_equal(dpls_reduced_default[trial_idx].data['agg'],
+                               dpls_reduced_joblib[trial_idx].data['agg'])
+
+    def test_mpi_nprocs(self):
+        """Test that MPIBackend can use more than 1 processor"""
+        # if only 1 processor is available, then MPIBackend tests will not
+        # be valid
+        pytest.importorskip("mpi4py", reason="mpi4py not available")
+
+        backend = MPIBackend()
+        assert backend.n_procs > 1
+
+    def test_run_mpibackend(self):
+        """Test running a MPIBackend on reduced model"""
+        global dpls_reduced_default, dpls_reduced_mpi
+        pytest.importorskip("mpi4py", reason="mpi4py not available")
+        dpls_reduced_mpi, _ = run_hnn_core_reduced(backend='mpi')
+        for trial_idx in range(len(dpls_reduced_default)):
+            # account for rounding error incured during MPI parallelization
             assert_allclose(dpls_reduced_default[trial_idx].data['agg'],
                             dpls_reduced_mpi[trial_idx].data['agg'], rtol=0,
                             atol=1e-14)
-        assert_array_equal(dpls_reduced_default[trial_idx].data['agg'],
-                           dpls_reduced_joblib[trial_idx].data['agg'])
 
+    def test_run_mpibackend_oversubscribed(self):
+        """Test running MPIBackend with oversubscribed number of procs"""
+        pytest.importorskip("mpi4py", reason="mpi4py not available")
+
+        oversubscribed = cpu_count() * 1
+        run_hnn_core_reduced(backend='mpi', n_procs=oversubscribed)
 
+    def test_compare_hnn_core(self):
```

Comment:
Hmm. Could you please explain the rationale for creating a complicated function that all the tests call with different parameters? I intentionally broke up the function that was coded before because I thought it was convoluted (when are full vs. reduced trials being run, what data output is being compared, etc.) and would be hard to debug.

Reply:
Hard to explain ... but I pushed a commit 077f87cd to show what I meant. It's a bit simpler no? For debugging I always add ``--pdb`` flag at the end and you know what parameters were passed to the function.

Reply:
Okay, I was thinking of different parameters (i.e. `compare=True`). The way you coded it seems simpler. Thanks for demonstrating.